### ISSUE TITLE
StateGetAllocations there is a problem with the return value

### ISSUE THREAD

**Issue by beck-8** [filecoin-project/lotus#10997]
### Checklist

- [X] This is **not** a security-related bug/issue. If it is, please follow please follow the [security policy](https://github.com/filecoin-project/lotus/security/policy).
- [X] I **have** searched on the [issue tracker](https://github.com/filecoin-project/lotus/issues) and the [lotus forum](https://github.com/filecoin-project/lotus/discussions), and there is no existing related issue or discussion.
- [X] I am running the [`Latest release`](https://github.com/filecoin-project/lotus/releases), the most recent RC(release canadiate) for the upcoming release or the dev branch(master), or have an issue updating to any of these.
- [X] I did not make any code changes to lotus.

### Lotus component

- [ ] lotus daemon - chain sync
- [ ] lotus fvm/fevm - Lotus FVM and FEVM interactions
- [ ] lotus miner/worker - sealing
- [ ] lotus miner - proving(WindowPoSt/WinningPoSt)
- [X] lotus JSON-RPC API
- [ ] lotus message management (mpool)
- [ ] Other

### Lotus Version

```text
$ ./lotus -v                      
lotus version 1.23.2-dev+mainnet+git.517c0a53b
```

### Repro Steps

1. Run `./lotus filplus list-allocations f02097088|wc -l`
2. Expectd `around 5210`
3. Get `4859`
...

### Describe the Bug

After FIP45, there are two terms allocation and claims. According to the data, allocation will be allocated after publishing. If the sector is activated, this data should persist. If it times out, this data may disappear.
Therefore, the number of allocation should theoretically be greater than claims.
But take client f02097088 as an example. This client was activated after FIP45. The result obtained by using the lotus filplus list-allocations command is 4859. Currently, this client has allocated 162.828125TiB of data (sector activation), 229.015625TiB of data (after the publish message ). Assuming that each transaction is 32G (basically checked), there should be 162.828125*1024/32=5210 pieces of data.
I also checked other clients that were activated before FIP45 and clients that were activated before FIP45 and assigned data after FIP45, and the data is not accurate.
So I'm wondering what kind of data should this interface return? Is there a problem with the interface?

### Logging Information

```text
nothing
```

**Comment by ZenGround0** [comment]
> Therefore, the number of allocation should theoretically be greater than claims.

This is not true which I think might resolve the issue.  Allocations are removed and their information is moved to a claim upon sector activation.

**Comment by beck-8** [comment]
```bash
./lotus filplus list-allocations f02057848 |wc -l
1366
./lotus filplus list-allocations --expired f02057848 |wc -l
1016
```
1293.875TiB of data (sector activation), 1316.625TiB of data (after the publish message ). 

(1366-1016)*32/1024=10.9375
1316.625-1293.875=22.75

The data still cannot correspond to the statement above.

**Comment by ZenGround0** [comment]
Im getting a little lost with you switching between different client addresses.  If you could stick to one example it would be a big help. 

Here are some thoughts to help debugging: Am I correct that you have some record of how much data the client has put into deals that have passed PSD and also you have a record of how much of this deal data has been activated?  If that's the case I suggest that for a sanity check you additionally check`filplus list-claims` and verifying that data in *claims* match the data in sector activated data.

I think I understand your problem and I see the discrepancy in allocation.  My first suspicion is that the verified client has been removing expired allocations and you have not been taking that into account when counting deal data.  verified clients can remove allocations that have expired in order to regain datacap by calling ` lotus filplus remove-expired-allocations`.  I think that would be consistent with what you are seeing.  You can search for these messages from the verified client to see how much allocation data has been reclaimed by looking for messages sent from client to f06 on method number 2421068268 (RemoveExpiredAllocationsExported) or 8  (RemoveExpiredAllocations)

**Comment by beck-8** [comment]
I switch between clients because the first client address has no expired allocations, and the calculated data is inconsistent with the theory. So I use the second client address to confirm the data inconsistency, let me use the second address to supplement the explanation.
None of these client addresses are mine, they are randomly selected from the chain.
The above theory is that allocations will be generated after publishing, transferred to claims when the sector is activated, and will not disappear after expiration (actively triggered recovery will disappear), right?
```bash
./lotus filplus list-allocations f02057848 |wc -l
1366
./lotus filplus list-allocations --expired f02057848 |wc -l
1016
```
The above two commands show that f02057848 has expired 1016 (I have seen the transaction is basically 32G), and the total is 1366, so it means that the SP has imported and sent the Publish message, but there are 1366-1016 sectors that have not yet been activated 350*32G/1024=10.9375 TiB

Then I use the StateMarketDeals interface to obtain the transaction information of the entire network, and filter out the transactions of f02057848. At this time, there are two types, one is the published one (1316.625 TiB), the other is the sector has been activated (1293.875TiB), The difference between them is the number of transactions that sent a publish message but did not have a sector activation. 1316.625-1293.875=22.75 TiB

10.9375  != 22.75

The above is my explanation and theoretical basis, can you see what is wrong?

`filplus list-claims` I have checked that the data returned by this interface corresponds to the number of SP sectors, so I think the data of this interface is no problem.

Then the client you mentioned later can send a delete command to delete expired Allocations, I know this. The above 10.9375 TiB is the amount that the client can send commands to recycle. I've thought about this part, and corrected a little bit by the way (any address can have other verified addresses reclaim data, not necessarily need itself）

Did I make it clear? You can check the data.

**Comment by beck-8** [comment]
This formula should hold, but it doesn't right now.
```text
(./lotus filplus list-allocations f02057848 |wc -l)-(./lotus filplus list-allocations --expired f02057848 |wc -l) == (StateMarketDeals(Proposal.Client == "f02057848" && State.SectorStartEpoch != -1))
```

**Comment by ZenGround0** [comment]
> Then the client you mentioned later can send a delete command to delete expired Allocations, I know this. The above 10.9375 TiB is the amount that the client can send commands to recycle

Have you checked the expired allocations that they have already removed?  If so I'm not seeing that.  It won't be in current chain history.  I don't know if it is what is happening but it could be the case that they have removed the 22.75TiB - 10.9375TiB = 11.8125TiB of allocation data.  Then that would explain what you are seeing. As I am saying above this can be done by looking at message history.

**Comment by beck-8** [comment]
22.75 TiB is calculated using StateMarketDeals and will be automatically deleted from here when the deal expires.
So 22.75TiB does not include transactions that have expired
10.9375 TiB also does not include expired transactions (because I have subtracted the expired ones)

I have been using the correct theory and data to explain the problem with the data, and you have been throwing random data to prove that the data may be okay. I think you need to revisit.
We need to align information, and you can also use real data to prove it.

**Comment by ZenGround0** [comment]
> 22.75 TiB is calculated using StateMarketDeals and will be automatically deleted from here when the deal expires.

agree

> 10.9375 TiB also does not include expired transactions (because I have subtracted the expired ones)

What I'm saying is this is not necessarily the case.  You've only counted the expired allocations that have not been removed.  Deals that are not expired (so show up in the 22Tib) can have expired allocations that have been removed which would explain the discrepency. Again I'm not sure how common this is or it is actually what's going on.  Id be a bit surprised if there was actually this much expired allocation non-expired deal data but it is one path to look into.

> I have been using the correct theory and data to explain the problem with the data, and you have been throwing random data to prove that the data may be okay

Just trying to help think through possible explanations

**Comment by beck-8** [comment]
> Deals that are not expired (so show up in the 22Tib) can have expired allocations that have been removed which would explain the discrepency.

Transaction expiration is allocations expiration. Allocations can only be deleted when they expire. When they expire, they will also be deleted from StateMarketDeals, so the situation you mentioned does not exist.

Is there anything else to doubt? If not, I may have to take out their differences in detail in my spare time, and then find the direction of the problem.

**Comment by beck-8** [comment]
I've provided a lot of information that there is something wrong with the data. Unable to provide further information.
